### ISSUE TITLE
Resize back buffer while window is being resized

### ISSUE THREAD

**Issue by Jjagg** [MonoGame/MonoGame#5590]
In XNA when the window is being resized by the user the back buffer is not resized. The changes are only pushed through when the resizing ends. MonoGame currently also does that for DX for XNA conformance. We don't do that for DesktopGL because SDL does not have an event for user resizing ending. For DesktopGL the back buffer is resized every frame while the user is resizing. I don't have a clue what UWP does honestly.

I opened this issue to propose the alternative to what XNA does. I.e. on all desktop platforms resize the back buffer every frame while the window is being resized. With current hardware this is really not an issue performance wise (which is maybe why XNA didn't do this). It looks neater (if handled correctly by the game dev, but they obviously should if they support resizing) and is easier to get working cross-platform in MonoGame itself. Thoughts?

cc @cra0zy

**Comment by KonajuGames** [comment]
I agree with that. It is not breaking behaviour to resize the back buffer
during the operation, and I think it looks cleaner. Performance is not an
issue from what I have seen.

**Comment by nkast** [comment]
UWP resizes the backbuffer while the user resizes the window. 
An interesting sideffect of the new gameloop and the way events are now synced was that the backbuffer is resized every frame to the last recorded event.

> With current hardware this is really not an issue performance wise (which is maybe why XNA didn't do this).
Most probably. It was the easier way to avoid what I described above. Just by subscribing to SizeChanged you will get multiple events between frames and you will unnecessary resize the backbuffer more that once before Update()/Draw() is called.

**Comment by tomspilman** [comment]
> In XNA when the window is being resized by the user the back buffer is not resized. The changes are only pushed through when the resizing ends.

I think we need to stick to this behavior.

Resizing the back buffer is often a new allocation...  doing 100s of allocations as you resize seems silly.  Not to mention the user could do a ton of allocations as well on the size changed events.

I think by default we should stretch the old back buffer to fit the new window size.  This gives the player a reasonable "preview" of what the game might look like after the resize is complete.

If we want to add a new feature to interactive resize of the back buffer we could do that.  It should likely be something people have to opt-in for.  That said... we would have added another test case for ourselves.

**Comment by cra0zy** [comment]
> Resizing the back buffer is often a new allocation... doing 100s of allocations as you resize seems silly. Not to mention the user could do a ton of allocations as well on the size changed events.
> 
> I think by default we should stretch the old back buffer to fit the new window size. This gives the player a reasonable "preview" of what the game might look like after the resize is complete.

XNA was designed for Windows, and before Windows didn't really do a redraw of the window until the resize ended (think of Windows XP resizing with outlines)... nowdays every window that gets resized updates its contents properly in real time so I don't think that calling that event only on resize end would be a correct move.

Also try dowinloading any UWP game and resizing their window, they will redraw stuff while they are being resized.

**Comment by nkast** [comment]
> Also try downloading any UWP game and resizing their window, they will redraw stuff while they are being resized.

We definitely need some test cases to know for sure, but here is my quick guess on why it behaves that way.

  On DesktopGL, where we resize on every event, it could slow down the entire main thread including the Windows drawing of the NC area (the windows borders, title bar, etc) making the app feel unresponsive.
  On UWP (or any XAML app) there are two threads (UI and Render). Recently MG also started using it's own thread for drawing. So I guess the only impact is a frame drop as the backbuffer is resized on every frame. The resizing of the window still feels responsive to the user.

**Comment by cra0zy** [comment]
I was not talking about MonoGame UWP game, I was just talking about any UWP game in general.

**Comment by nkast** [comment]
And I was talking about XAML apps in general! (including UWP)

My point is, whether MG's Draw(), Win32's OnPaint(), or XAML's Composition blocks the window message queue or not.

**Comment by mrhelmut** [comment]
Considering how Mac OS handles backbuffer in Retina mode, resizing the backbuffer seems to be the most compatible alternative accross platforms. Mac OS really doesn't like that the backbuffer is not scaled properly with the window.

**Comment by Apostolique** [comment]
Any chance we can at least add an opt in option? Would be really nice to continuously resize the back buffer for my game.

**Comment by nkast** [comment]
@Apostolique 
You don't need a specific event for that. You can update the backbuffer on every Draw() from `Window.ClientBounds`.

My understanding is that `Window.ClientBounds` is not implemented the same way in every platform.
I don't know on which platform it work if any, I just tested it on an XNA app.

**Comment by Apostolique** [comment]
The value for Window.ClientBounds only gets updated after the resizing is done. I'm testing in Windows 10.

**Comment by nkast** [comment]
you mean Windows 10 store App? 

I see you already have an open issue for OpenGL #5664 as well.

**Comment by Apostolique** [comment]
Sorry, didn't mean Windows 10 Store. Just regular Windows platform. But as you noticed, my project is DesktopGL.

**Comment by nkast** [comment]
That's strange. 
OpenGL resizes the backbuffer while the user resizes the window but updates ClientsBounts at the end of resizing?

**Comment by Apostolique** [comment]
This is still one the main issue I have with 3.6 and 3.7.
